### PR TITLE
#204 Blob recovery tool

### LINKED ISSUE #204: Implement tool to extract data from corrupted BLOB
Clone of [this](https://github.com/qoollo/pearl/issues/87) issue

### PR DESCRIPTION
**Pull request by vovac12**
Closes #204

### CODE REVIEW

File: bob-tools/bin/brt.rs [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    let mut output = Blob::writer(&settings.output)?;
+    log::info!("Blob writer created");
+    let header = input.read_header()?;
+    output.write_header(&header)?;
+    log::info!("Input blob header version: {}", header.version);
+    let mut count = 0;
+    while !input.is_eof() {
+        match input.read_record() {
+            Ok(record) => {
+                output.write_record(&record)?;
```
Comment (idruzhitskiy) [inline bob-tools/bin/brt.rs:207]: Maybe there should be some checksum validation mid-write, like every N records we check that checksum of written records are the same as the read ones (so we keep at max N records in memory, N can be read from config with default value of 100, for example).

Comment (vovac12) [reply bob-tools/bin/brt.rs]: Now I check the equality of N last cached and written records.

File: bob-tools/bin/brt.rs [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+impl Header {
+    fn validate(&self) -> AnyResult<()> {
+        if self.magic_byte != RECORD_MAGIC_BYTE {
+            return Err(ValidationError("record header magic byte is invalid".to_string()).into());
+        }
+        let mut header = self.clone();
+        header.header_checksum = 0;
+        let serialized = bincode::serialize(&header)?;
+        let checksum = crc32(&serialized);
```
[the remark this answers is not in the document]
Comment (vovac12) [reply bob-tools/bin/brt.rs]: I created a separate function to check the checksum of a vector of bytes.

File: bob-tools/bin/brt.rs [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+impl Error for ValidationError {}
+
+struct Blob {
+    file: File,
+    position: u64,
+    len: u64,
+}
+
+impl Blob {
```
[the remark this answers is not in the document]
Comment (vovac12) [reply bob-tools/bin/brt.rs]: I created 2 separate structures

**Comment by idruzhitskiy** [comment]
LGTM

**Review by pyakushin — APPROVED** [review]
LGTM, fix conflicts in readme

**Comment by vovac12** [comment]
Fixed

**Pull request merged**